matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -229,52 +231,77 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
+        ViterbiAlgorithm<GPXExtension, GPXEntry, Path> viterbi = new ViterbiAlgorithm<>();
+        final List<List<SequenceState<GPXExtension, GPXEntry, Path>>> sequences =
+                new ArrayList<List<SequenceState<GPXExtension, GPXEntry, Path>>>();
+        TimeStep<GPXExtension, GPXEntry, Path> seqPrevTimeStep = null;
         for (TimeStep<GPXExtension, GPXEntry, Path> timeStep : timeSteps) {
+
+            // if sequence is broken, then close it off and create a new viterbi:
+            if (viterbi.isBroken()) {
+                sequences.add(viterbi.computeMostLikelySequence());
+                seqPrevTimeStep = null;
+                viterbi = new ViterbiAlgorithm<>();
+            }
+            
+            // always calculate emission probabilities regardless of place in sequence:
             computeEmissionProbabilities(timeStep, probabilities);
 
-            if (prevTimeStep == null) {
+            if (seqPrevTimeStep == null) {
+                // first step of a sequence, so initialise viterbi:
                 viterbi.startWithInitialObservation(timeStep.observation, timeStep.candidates,
                         timeStep.emissionLogProbabilities);
+                // it is possible viterbi is immediately broken here (e.g. no candidates) - this
+                // will be caught by the first test in this loop.
             } else {
-                computeTransitionProbabilities(prevTimeStep, timeStep, probabilities, queryGraph);
+                // add this step to current sequence:
+                computeTransitionProbabilities(seqPrevTimeStep, timeStep, probabilities, queryGraph);
                 viterbi.nextStep(timeStep.observation, timeStep.candidates,
                         timeStep.emissionLogProbabilities, timeStep.transitionLogProbabilities,
                         timeStep.roadPaths);
-            }
-            if (viterbi.isBroken()) {
-                String likelyReasonStr = "";
-                if (prevTimeStep != null) {
-                    GPXEntry prevGPXE = prevTimeStep.observation;
-                    GPXEntry gpxe = timeStep.observation;
-                    double dist = distanceCalc.calcDist(prevGPXE.lat, prevGPXE.lon,
-                            gpxe.lat, gpxe.lon);
-                    if (dist > 2000) {
-                        likelyReasonStr = "Too long distance to previous measurement? "
-                                + Math.round(dist) + "m, ";
-                    }
+                // if broken, then close off this sequence and create a new one starting with this
+                // timestep. Note that we rely on the fact that if the viterbi breaks the most
+                // recent step does not get added i.e. 'computeMostLikelySequence' returns the most
+                // likely sequence without this (breaking) step added. Hence we can use it to start
+                // the next one:
+                // TODO: check the above is true
```

Comment:
@stefanholder - I think this is true, but can you confirm? I guess I could check for it to some degree ...